File: C++/Searching and Sorting/29_max_sum_path.cpp
/*
Idea:
using 2 arrays holding sum of elements till the pointers
-at intersection (no. at indices being equal) , changing sum of both arrays to the max
-if any array ends, adding rest of elements till both ends.
returning max value
*/

#include<bits/stdc++.h>
using namespace std;



class Solution
{
public:
    int max_path_sum(int A[], int B[], int l1, int l2)
    {
        int sum1 = 0, sum2 = 0, i = 0, j = 0;
        while (i < l1 && j < l2)
        {
            if (A[i] == B[j])
            {
                sum1 = sum2 = max(sum1, sum2) + A[i];
                ++i;
                ++j;
            }
            else if (A[i] < B[j])
                sum1 += A[i++];
            else
                sum2 += B[j++];
        }

        while (i < l1)
            sum1 += A[i++];
        while (j < l2)
            sum2 += B[j++];
        return max(sum1, sum2);


    }
};


int main()
{
    int T;
    cin >> T;

    while (T--)
    {
        int N, M;
        cin >> N >> M;
        fflush(stdin);
        int a[N], b[M];
        for (int i = 0;i < N;i++)
            cin >> a[i];
        for (int i = 0;i < M;i++)
            cin >> b[i];
        Solution obj;
        int result = obj.max_path_sum(a, b, N, M);
        cout << result << endl;
    }
}

File: Python/Searching and Sorting/29_max_sum_path.py
"""
Idea:
using 2 arrays holding sum of elements till the pointers
-at intersection (no. at indices being equal) , changing sum of both arrays to the max
-if any array ends, adding rest of elements till both ends.
returning max value
"""


class Solution:
    def maxSumPath(self, arr1, arr2, m, n):
        sum1 = sum2 = 0
        i = j = 0
        while i < m and j < n:
            if arr1[i] == arr2[j]:
                sum1 = sum2 = max(sum1, sum2)+arr1[i]
                i += 1
                j += 1
            elif arr1[i] < arr2[j]:
                sum1 += arr1[i]
                i += 1
            else:
                sum2 += arr2[j]
                j += 1
        while i < m:
            sum1 += arr1[i]
            i += 1
        while j < n:
            sum2 += arr2[j]
            j += 1
        return max(sum1, sum2)


if __name__ == '__main__':
    t = int(input())
    for i in range(t):
        m, n = list(map(int, input().strip().split()))
        arr1 = list(map(int, input().strip().split()))
        arr2 = list(map(int, input().strip().split()))
        print(Solution().maxSumPath(arr1, arr2, m, n))
